File: skia-rs-wasm/src/lib/renderer/api/material.ts
/**
 * Custom SkSL material (shader fill) operations.
 *
 * A material is a user-authored SkSL shader bound as a single shape's fill.
 * Its uniforms are reflected/bound by name on the Rust side, so the payload
 * just carries the source plus the current uniform values.
 *
 * Shared-memory layout matches `wasm/material.rs::parse_material` (little-endian):
 *   [u32 hidden]
 *   [u32 source_len] [source bytes] [pad → 4B]
 *   [u32 uniform_count]
 *   per uniform: [u32 name_len] [name bytes] [pad → 4B] [u32 comp] [comp × f32]
 * where `comp` (1..=4) is the component count → f32/vec2/vec3/vec4.
 */

import type { WasmModule } from '../wasm-types'
import { checkContext } from './context'
import { allocBytes, freeBytes } from '../utils'

export type MaterialUniformValue =
  | { type: 'f32'; value: number }
  | { type: 'vec2'; value: readonly [number, number] }
  | { type: 'vec3'; value: readonly [number, number, number] }
  | { type: 'vec4'; value: readonly [number, number, number, number] }

export interface MaterialUniform {
  name: string
  value: MaterialUniformValue
}

export interface Material {
  /** SkSL source, compiled to this shape's fill shader. */
  source: string
  /** Current uniform values, matched to the shader's uniforms by name. */
  uniforms?: MaterialUniform[]
  hidden?: boolean
}

/** A uniform reflected from compiled SkSL — drives which control the editor shows. */
export interface ReflectedUniform {
  name: string
  /** Component count 1..4; `0` for types with no simple control (matrices). */
  components: number
  /** True when declared `layout(color)` — render a color swatch, not sliders. */
  isColor: boolean
  /** Array length (1 when scalar). */
  count: number
}

/** Result of compiling a material's source: status + reflected uniforms/inputs. */
export interface MaterialCompileResult {
  ok: boolean
  /** Compile error message when `ok` is false. */
  error?: string
  uniforms: ReflectedUniform[]
  /** `uniform shader` input names (e.g. `content`/`backdrop`/`field`). */
  inputs: string[]
  /**
   * True when the source declares `u_time` — the material is clock-driven, so
   * the editor offers transport and runs an animation loop. `u_time` is
   * engine-owned and never appears in `uniforms`, so this is the only signal.
   */
  usesTime: boolean
}

const COMP_COUNT: Record<MaterialUniformValue['type'], number> = {
  f32: 1,
  vec2: 2,
  vec3: 3,
  vec4: 4,
}

const align4 = (n: number): number => (n + 3) & ~3

/**
 * Stage a material into shared memory in the LE layout above. The caller then
 * invokes the matching wasm entry point and calls `freeBytes`.
 *
 * Shared by the on-canvas path (`_set_shape_material`) and the isolated focus
 * preview (`_preview_set_material`) so the layout is written in exactly one
 * place and the two can't drift.
 */
function stageMaterialPayload(module: WasmModule, material: Material): void {
  const enc = new TextEncoder()
  const source = enc.encode(material.source)
  const uniforms = material.uniforms ?? []
  const names = uniforms.map((u) => enc.encode(u.name))

  let size = 4 // hidden
  size += 4 + align4(source.length) // source_len + source + pad
  size += 4 // uniform_count
  for (let i = 0; i < uniforms.length; i++) {
    size += 4 + align4(names[i].length) // name_len + name + pad
    size += 4 + COMP_COUNT[uniforms[i].value.type] * 4 // comp + values
  }

  const offset = allocBytes(module, size)
  const heap = module.HEAPU8
  const dv = new DataView(heap.buffer, heap.byteOffset)
  let p = offset

  dv.setUint32(p, material.hidden ? 1 : 0, true)
  p += 4
  dv.setUint32(p, source.length, true)
  p += 4
  heap.set(source, p)
  p += align4(source.length)
  dv.setUint32(p, uniforms.length, true)
  p += 4

  for (let i = 0; i < uniforms.length; i++) {
    const name = names[i]
    dv.setUint32(p, name.length, true)
    p += 4
    heap.set(name, p)
    p += align4(name.length)

    const v = uniforms[i].value
    dv.setUint32(p, COMP_COUNT[v.type], true)
    p += 4
    if (v.type === 'f32') {
      dv.setFloat32(p, v.value, true)
      p += 4
    } else {
      for (let k = 0; k < v.value.length; k++) {
        dv.setFloat32(p, v.value[k], true)
        p += 4
      }
    }
  }
}

/**
 * Set the custom SkSL material on the current shape. Clears when `material`
 * is null/undefined or has no source.
 */
export function setShapeMaterial(module: WasmModule, material: Material | null | undefined): void {
  checkContext()
  if (!material || !material.source) {
    module._clear_shape_material()
    return
  }
  stageMaterialPayload(module, material)
  module._set_shape_material()
  freeBytes(module)
}

/**
 * Set the material rendered by the isolated focus preview. Same payload as
 * `setShapeMaterial`, different entry point — the preview keeps its own copy
 * and never touches the document's shape tree.
 *
 * Only stages + parses (no GL work), so it doesn't require the preview context
 * to be current — but `focus-preview.ts` calls it inside the same make-current
 * block as the draw anyway, which is harmless and keeps the sequence obvious.
 */
export function setPreviewMaterial(module: WasmModule, material: Material | null | undefined): void {
  if (typeof module._preview_set_material !== 'function') return
  if (!material || !material.source) {
    module._preview_clear_material()
    return
  }
  stageMaterialPayload(module, material)
  module._preview_set_material()
  freeBytes(module)
}

/**
 * Compile a material's SkSL source and reflect its editable uniforms + shader
 * inputs. Mirrors `wasm/material.rs::compile_material`'s result layout (LE):
 *   [u32 ok][u32 err_len][err bytes][u32 uniform_count]
 *   per uniform: [u32 name_len][name][u32 components][u32 is_color][u32 count]
 *   [u32 input_count] per input: [u32 name_len][name]
 *
 * The editor uses this to build controls and surface compile errors.
 */
export function compileMaterial(module: WasmModule, source: string): MaterialCompileResult {
  checkContext()

  // Guard against a stale binary that predates this export: calling a missing
  // function throws *after* we've staged the input buffer, leaking it and
  // cascading into a "Bytes already allocated" panic on the next alloc.
  if (typeof module._compile_material !== 'function') {
    return {
      ok: false,
      error: 'Renderer is out of date — rebuild the WASM (pnpm --filter skia-rs-wasm build:wasm).',
      uniforms: [],
      inputs: [],
      usesTime: false,
    }
  }

  const src = new TextEncoder().encode(source)
  const inOffset = allocBytes(module, Math.max(src.length, 1))
  try {
    module.HEAPU8.set(src, inOffset)

    const ptr = module._compile_material()

    // Re-read the heap after the call — wasm memory may have grown, invalidating
    // any earlier HEAPU8 view.
    const heap = module.HEAPU8
    const dv = new DataView(heap.buffer, heap.byteOffset)
    const dec = new TextDecoder()
    let p = ptr

    const readU32 = (): number => {
      const v = dv.getUint32(p, true)
      p += 4
      return v
    }
    const readStr = (): string => {
      const n = readU32()
      const s = dec.decode(heap.subarray(p, p + n))
      p += n
      return s
    }

    const ok = readU32() === 1
    const error = readStr()

    const uniformCount = readU32()
    const uniforms: ReflectedUniform[] = []
    for (let i = 0; i < uniformCount; i++) {
      const name = readStr()
      const components = readU32()
      const isColor = readU32() === 1
      const count = readU32()
      uniforms.push({ name, components, isColor, count })
    }

    const inputCount = readU32()
    const inputs: string[] = []
    for (let i = 0; i < inputCount; i++) {
      inputs.push(readStr())
    }
    const usesTime = readU32() === 1

    return {
      ok,
      error: error.length > 0 ? error : undefined,
      uniforms,
      inputs,
      usesTime,
    }
  } finally {
    // Always release the staged buffer — even if parsing throws — so a failure
    // can't leave BUFFERU8 allocated and panic the next alloc.
    freeBytes(module)
  }
}

File: render-wasm/src/render/ssa/material.rs
//! SSA renderer for a single-shape custom SkSL material — a post-body
//! overlay effect (drawn over the shape's fill/strokes, not replacing them).
//!
//! Unlike `glass`/`noise` — whose uniforms are hand-matched by name — a
//! material's uniforms are bound generically: the binder walks
//! `effect.uniforms()` and, for each declared uniform, either fills an
//! engine-owned value (`u_resolution`, `u_scale`, `u_time`) or copies the
//! matching user value from `material.uniforms`. Adding a `uniform` to the
//! source therefore needs no Rust change.
//!
//! Scope (first slice): self-contained shaders only — no child samplers
//! (`backdrop`/`field`) and `u_time` is always `0.0` (no clock yet). The
//! shader is evaluated in shape-local coordinates: `fragCoord` spans
//! `0..selrect.size` and `u_resolution` is that size.

use std::cell::RefCell;
use std::collections::hash_map::DefaultHasher;
use std::collections::HashMap;
use std::hash::{Hash, Hasher};

use skia_safe::{self as skia, RuntimeEffect};

use crate::error::Result;
use crate::shapes::{Material, Shape, UniformValue};

use super::PaintCtx;

/// Maximum device-pixel extent of a material's render buffer. Beyond this the
/// shader is rendered into a reduced offscreen and upsampled, so extreme zoom
/// can't make the fragment shader run over an unbounded pixel count (or exceed
/// the ~4096 minimum WebGL max-texture-size). Material patterns are
/// low-frequency, so the upscale is visually cheap.
const RES_CAP_PX: f32 = 2048.0;

/// Uniforms the renderer fills itself — never surfaced as editor controls.
const ENGINE_UNIFORMS: [&str; 3] = ["u_resolution", "u_scale", "u_time"];

/// Max compiled effects retained. The cache is a *liveness memo*, not a
/// history: live sources are re-requested every frame so they stay resident,
/// while dead ones (every keystroke in the editor produces a source that is
/// compiled once and never looked up again) age out. Without a bound this
/// grows forever — a real leak once the editor/preview compile per keystroke.
const CACHE_CAP: usize = 64;

thread_local! {
    /// Compiled effects keyed by source hash. User sources aren't `const`,
    /// so a single `OnceCell` (as glass/noise use) won't do — distinct
    /// sources must coexist, and an unchanged source must hit the cache.
    ///
    /// LRU-bounded to `CACHE_CAP`: `order` holds keys oldest-first and is
    /// touched on every hit. Eviction only costs a recompile, never
    /// correctness.
    static CACHE: RefCell<HashMap<u64, RuntimeEffect>> = RefCell::new(HashMap::new());
    static CACHE_ORDER: RefCell<Vec<u64>> = const { RefCell::new(Vec::new()) };
}

/// Mark `key` as most-recently-used, evicting the oldest entries past the cap.
fn touch_cache_key(key: u64) {
    CACHE_ORDER.with(|order| {
        let mut order = order.borrow_mut();
        if let Some(pos) = order.iter().position(|k| *k == key) {
            order.remove(pos);
        }
        order.push(key);
        while order.len() > CACHE_CAP {
            let evicted = order.remove(0);
            CACHE.with(|cache| {
                cache.borrow_mut().remove(&evicted);
            });
        }
    });
}

/// Engine-supplied uniform values. Never user-set; filled by the renderer.
/// `pub(crate)` so the isolated focus-mode preview (`render::preview`) can bind
/// the same uniforms against its own surface instead of duplicating them.
pub(crate) struct EngineUniforms {
    pub resolution: (f32, f32),
    pub scale: f32,
    pub time: f32,
}

fn hash_source(src: &str) -> u64 {
    let mut h = DefaultHasher::new();
    src.hash(&mut h);
    h.finish()
}

/// Compile (cached). Returns the cached effect on hit; compiles + caches on
/// miss. Never panics — a compile error is returned as `Err(message)` so the
/// caller can keep the last good frame instead of crashing.
fn get_or_compile(src: &str) -> std::result::Result<RuntimeEffect, String> {
    let key = hash_source(src);
    let hit = CACHE.with(|cache| cache.borrow().get(&key).cloned());
    if let Some(effect) = hit {
        touch_cache_key(key);
        return Ok(effect);
    }
    let effect = RuntimeEffect::make_for_shader(src, None).map_err(|e| e.to_string())?;
    CACHE.with(|cache| {
        cache.borrow_mut().insert(key, effect.clone());
    });
    // Insert then evict, so `touch` can drop the oldest without ever evicting
    // the entry we just added (it's the newest).
    touch_cache_key(key);
    Ok(effect)
}

/// A reflected uniform surfaced to the editor so it can auto-generate a control
/// (a slider per component, or a color swatch when `is_color`).
pub struct ReflectedUniform {
    pub name: String,
    /// 1..=4; `0` for types with no simple control (matrices).
    pub components: u32,
    pub is_color: bool,
    /// Array length (1 when scalar).
    pub count: u32,
}

/// Outcome of compiling a material's source for the editor: compile status plus
/// the reflected editable uniforms and `uniform shader` inputs.
pub struct ReflectResult {
    pub ok: bool,
    pub error: Option<String>,
    pub uniforms: Vec<ReflectedUniform>,
    pub inputs: Vec<String>,
}

fn type_components(ty: skia::runtime_effect::uniform::Type) -> u32 {
    use skia::runtime_effect::uniform::Type;
    match ty {
        Type::Float | Type::Int => 1,
        Type::Float2 | Type::Int2 => 2,
        Type::Float3 | Type::Int3 => 3,
        Type::Float4 | Type::Int4 => 4,
        _ => 0,
    }
}

/// Compile (cached) and reflect the material's editable uniforms + shader
/// inputs. Engine-owned uniforms are omitted. On compile error, `ok = false`
/// carries the message and the lists are empty.
pub fn compile_and_reflect(src: &str) -> ReflectResult {
    let effect = match get_or_compile(src) {
        Ok(e) => e,
        Err(error) => {
            return ReflectResult {
                ok: false,
                error: Some(error),
                uniforms: Vec::new(),
                inputs: Vec::new(),
            }
        }
    };
    let uniforms = effect
        .uniforms()
        .iter()
        .filter(|u| !ENGINE_UNIFORMS.iter().any(|e| *e == u.name()))
        .map(|u| ReflectedUniform {
            name: u.name().to_string(),
            components: type_components(u.ty()),
            is_color: u.is_color(),
            count: u.count().max(1) as u32,
        })
        .collect();
    let inputs = effect
        .children()
        .iter()
        .map(|c| c.name().to_string())
        .collect();
    ReflectResult {
        ok: true,
        error: None,
        uniforms,
        inputs,
    }
}

fn write_f32(data: &mut [u8], offset: usize, val: f32) {
    data[offset..offset + 4].copy_from_slice(&val.to_ne_bytes());
}

fn write_value(data: &mut [u8], offset: usize, value: &UniformValue) {
    match value {
        UniformValue::F32(x) => write_f32(data, offset, *x),
        UniformValue::Vec2(v) => v.iter().enumerate().for_each(|(i, x)| write_f32(data, offset + i * 4, *x)),
        UniformValue::Vec3(v) => v.iter().enumerate().for_each(|(i, x)| write_f32(data, offset + i * 4, *x)),
        UniformValue::Vec4(v) => v.iter().enumerate().for_each(|(i, x)| write_f32(data, offset + i * 4, *x)),
    }
}

/// Pack the uniform buffer: engine-owned uniforms by fixed name, everything
/// else from the material's slots matched by name. Unmatched uniforms stay
/// zeroed.
fn bind(effect: &RuntimeEffect, material: &Material, engine: &EngineUniforms) -> Vec<u8> {
    let mut data = vec![0u8; effect.uniform_size()];
    for u in effect.uniforms().iter() {
        let name: &str = &u.name();
        let off = u.offset();
        match name {
            "u_resolution" => {
                write_f32(&mut data, off, engine.resolution.0);
                write_f32(&mut data, off + 4, engine.resolution.1);
            }
            "u_scale" => write_f32(&mut data, off, engine.scale),
            "u_time" => write_f32(&mut data, off, engine.time),
            _ => {
                if let Some(slot) = material.uniforms.iter().find(|s| s.name == name) {
                    write_value(&mut data, off, &slot.value);
                }
            }
        }
    }
    data
}

/// Compile + bind a material into a paint shader. `None` on compile error
/// (caller draws nothing rather than crashing). No child samplers yet.
///
/// `pub(crate)` so the isolated preview reuses the exact same compile + bind
/// path as the on-canvas render — the preview can't drift from the real thing.
pub(crate) fn make_material_shader(
    material: &Material,
    engine: &EngineUniforms,
    local_matrix: Option<&skia::Matrix>,
) -> Option<skia::Shader> {
    let effect = get_or_compile(&material.source).ok()?;
    let data = bind(&effect, material, engine);
    let children: [skia::runtime_effect::ChildPtr; 0] = [];
    effect.make_shader(skia::Data::new_copy(&data), &children, local_matrix)
}

/// Draw the material shader over the shape's finished body, clipped to the
/// shape's geometry and composited SrcOver (non-destructive — the fill and
/// strokes underneath show through wherever the shader outputs alpha < 1).
///
/// The shader is evaluated in shape-local world units (`u_resolution =
/// selrect.size`, `fragCoord` spans `0..size`). When the shape's on-screen
/// footprint exceeds `RES_CAP_PX` the shader is rendered into a reduced
/// offscreen and upsampled — bounding fragment work at extreme zoom.
pub fn render(ctx: &mut PaintCtx<'_>, shape: &Shape, material: &Material) -> Result<()> {
    let selrect = shape.selrect;
    let (w, h) = (selrect.width(), selrect.height());
    if w <= 0.0 || h <= 0.0 {
        return Ok(());
    }

    // Device-pixel footprint at the current zoom, and the cap factor `q <= 1`
    // that keeps the render buffer's longest side within `RES_CAP_PX`.
    let (dw, dh) = (w * ctx.scale, h * ctx.scale);
    let longest = dw.max(dh);
    let q = if longest > RES_CAP_PX {
        RES_CAP_PX / longest
    } else {
        1.0
    };

    // `u_resolution` stays in world units (zoom-independent pattern); `u_scale`
    // reflects the scale of the buffer actually being rendered into.
    let engine = EngineUniforms {
        resolution: (w, h),
        scale: ctx.scale * q,
        time: 0.0,
    };
    // Offset the shader so `fragCoord` starts at the shape's top-left.
    let local = skia::Matrix::translate((selrect.x(), selrect.y()));
    let Some(shader) = make_material_shader(material, &engine, Some(&local)) else {
        return Ok(());
    };

    if q >= 1.0 {
        // Footprint within budget — draw the shader straight onto the target.
        // Compute the tile+shape matrix before borrowing the canvas (`&self`
        // method vs the `&mut surface` borrow below).
        let matrix = ctx.tile_and_shape_transform_matrix(shape);
        let canvas = ctx.surface.canvas();
        canvas.save();
        canvas.concat(&matrix);
        crate::render::glass::clip_to_shape(canvas, shape);
        let mut paint = skia::Paint::default();
        paint.set_anti_alias(true);
        paint.set_shader(shader);
        canvas.draw_paint(&paint);
        canvas.restore();
        return Ok(());
    }

    // Over budget — render the shader into a capped offscreen, then upsample.
    let ow = ((dw * q).round() as i32).max(1);
    let oh = ((dh * q).round() as i32).max(1);
    let mut off = ctx.gpu.create_budgeted_surface(ow, oh)?;
    {
        let oc = off.canvas();
        oc.clear(skia::Color::TRANSPARENT);
        // Map the shape's world rect onto the offscreen's pixels.
        oc.scale((ow as f32 / w, oh as f32 / h));
        oc.translate((-selrect.x(), -selrect.y()));
        crate::render::glass::clip_to_shape(oc, shape);
        let mut paint = skia::Paint::default();
        paint.set_anti_alias(true);
        paint.set_shader(shader);
        oc.draw_paint(&paint);
    }
    let image = off.image_snapshot();

    let matrix = ctx.tile_and_shape_transform_matrix(shape);
    let sampling = ctx.sampling;
    let canvas = ctx.surface.canvas();
    canvas.save();
    canvas.concat(&matrix);
    crate::render::glass::clip_to_shape(canvas, shape);
    let src = skia::Rect::new(0.0, 0.0, ow as f32, oh as f32);
    let paint = skia::Paint::default();
    canvas.draw_image_rect_with_sampling_options(
        &image,
        Some((&src, skia::canvas::SrcRectConstraint::Fast)),
        selrect,
        sampling,
        &paint,
    );
    canvas.restore();
    Ok(())
}
